### PR TITLE
feat: proof of concept wasm safe_exec

### PR DESCRIPTION
**Pull request by ormsbee**
This super-hacky, just to prove it can be done. Works with very simple customresponse code that just uses standard library functions, and not numpy/scipy/etc.

`json_safe` was lifted from codejail

Most of the interesting changes are in `untrusted_code` and `safe_exec.py`

### CODE REVIEW

File: openedx-wasm/runner.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+from wasmtime import Config, Engine, Linker, Module, Store, WasiConfig
```
Comment (ormsbee) [inline openedx-wasm/runner.py:1]: Just ignore this file for the sake of this PR–it was mostly copy-pasted from an example gist someone had put up, and I used it for my very first attempts to start the wasm python.

**Comment by bradenmacdonald** [comment]
Thanks so much for this! Otherwise I wasn't able to test python-containing problems on my Mac.

To get this working, I had to:
1. Download https://github.com/vmware-labs/webassembly-language-runtimes/releases/download/python%2F3.11.3%2B20230428-7d1b259/python-3.11.3.wasm into `edx-platform/openedx-wasm/python/python-3.11.3/bin`
2. `mkdir openedx-wasm/logs/` (since the folder doesn't exist, wasmtime can't create the logfile).

Once I figured those out, it worked right away.

File: openedx-wasm/untrusted_code/run.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+pct_a_str = stringify(pct_a*100, False)
+pct_b_str = stringify(pct_b*100, False)
+beg_ada_str = stringify(beg_ada)
+end_ada_str = stringify(end_ada)
+writeoffs_str = stringify(writeoffs)   
+bde_str = stringify(bde)
+
+
+
+wasm_helper.write_globals_to_file(globals())
```
Comment (bradenmacdonald) [inline openedx-wasm/untrusted_code/run.py:45]: Seems like this file or perhaps everything in `untrusted_code` should be .gitignored?

File: .gitignore [hunk 3]
```diff
@@ -149,3 +149,7 @@ pii_report
 
 # Pyenv Python version file
 .python-version
+
+openedx-wasm/untrusted/*.json
+openedx-wasm/untrusted/run.py
```
Comment (bradenmacdonald) [inline .gitignore]: Needs to be `untrusted_code` I think, not `untrusted`

**Comment by ormsbee** [comment]
@bradenmacdonald: Where do you want to try to go with this next? I feel like there are broadly two areas we can expand in:

1. Figuring out how to properly build the third-party libraries we need to actually get to parity with what's used today.
2. Building out the interface for what the "real" version of this looks like–which I'm guessing is a new library in a separate repo that can fulfill the basic things safe_exec asks for.

Thoughts?

**Comment by bradenmacdonald** [comment]
I'm more interested in 2 for now, especially if it means that Tutor users can get a working sandboxed python capa problem that works out of the box, without needing a codejail plugin etc. That is already a nice improvement and we know that the tech is already there thanks to your proof of concept.

For 1, getting the full science stack, it seems like that's going to be hard today but may be easier tomorrow, so it makes sense to me if we tackle that one a bit further down the road.

**Comment by ormsbee** [comment]
@bradenmacdonald: Yeah, I agree that (2) makes more sense for now. If you code up a plugin, please tag me. 😄

**Comment by ormsbee** [comment]
Note that this approach is very outdated, and I would absolutely use the Wasm Component Model and [componentize-py](https://component-model.bytecodealliance.org/language-support/python.html#building-a-component-with-componentize-py) in the next attempt.

**Pull request closed without merging**